File: leet_code/linked_lists/merge_k_lists.js
// Merge k sorted linked lists and return it as one sorted list.
//
// Example:
//
// Input:
// [
//   1->4->5,
//   1->3->4,
//   2->6
// ]
// Output: 1->1->2->3->4->4->5->6


const mergeKLists = arr => {
  let newHead = null,
    cur = null;

  while (arr.length > 0) {
    let nextVal = null,
      idx = 0;

    for (let i = 0; i < arr.length; i++) {
      if (arr[i] === null) {
        continue;
      }
      if (nextVal === null || arr[i].val < nextVal) {
        nextVal = arr[i].val;
        idx = i;
      }
    }
    if (arr[idx] == null ) {
      arr = arr.slice(0,idx).concat(arr.slice(idx+1, arr.length));
      continue;
    } else if (newHead === null) {
      newHead = new ListNode(nextVal);
      cur = newHead;
    } else {
      cur.next = new ListNode(nextVal);
      cur = cur.next;
    }
    arr[idx] = arr[idx].next;
  }
  return newHead;
};

File: leet_code/linked_lists/merge_k_lists.rb
# // Merge k sorted linked lists and return it as one sorted list.
# //
# // Example:
# //
# // Input:
# // [
# //   1->4->5,
# //   1->3->4,
# //   2->6
# // ]
# // Output: 1->1->2->3->4->4->5->6
#
#

def merge_k_lists(arr)
  new_head, cur = null, null

  until arr.empty?
    next_val, idx = nil, 0

    arr.each_with_index do |node, i|
      next if node == nil
      if next_val == nil || node.val < next_val
        next_val = node.val
        idx = i
      end
    end

    if arr[idx] == nil
      arr.delete_at(idx)
      next
    elsif new_head == nil
      new_head = ListNode.new(next_val)
      cur = new_head
    else
      cur.next = ListNode.new(next_val)
      cur = cur.next
    end
    arr[idx] = arr[idx].next
  end
  new_head
end
